buffer/src/main/java/io/netty/buffer/PoolThreadCache.java

```diff
@@ -120,9 +114,22 @@ public void run() {
             numShiftsNormalHeap = -1;
         }
 
-        // The thread-local cache will keep a list of pooled buffers which must be returned to
-        // the pool when the thread is not alive anymore.
-        ThreadDeathWatcher.watch(thread, freeTask);
+        // We only need to watch the thread when any cache is used.
+        if (tinySubPageDirectCaches != null || smallSubPageDirectCaches != null || normalDirectCaches != null
+                || tinySubPageHeapCaches != null || smallSubPageHeapCaches != null || normalHeapCaches != null) {
+            freeTask = new Runnable() {
+                @Override
+                public void run() {
+                    free0(Thread.currentThread());
+                }
+            };
+
+            // The thread-local cache will keep a list of pooled buffers which must be returned to
+            // the pool when the thread is not alive anymore.
+            ThreadDeathWatcher.watch(Thread.currentThread(), freeTask);
```

Comment:
don't we still want to be sure we use the same thread when we call `watch` and `unwatch`? consider keeping the `Thread` local variable and re-name to `deathWatchThread` to clarify.